ktor-http/common/src/io/ktor/http/Url.kt

```diff
@@ -36,6 +36,109 @@ public class Url internal constructor(
+     *
+     * ```kotlin
+     * val fullUrl = Url("http://ktor.io/docs/")
+     * fullUrl.segments ==  listOf("docs")
+     *
+     * val absolute = Url("/docs/")
+     * absolute.segments ==  listOf("docs")
+     *
+     * val relative = Url("docs")
+     * relative.segments = listOf("docs")
+     * ```
+     *
+     * To address this issue, current [pathSegments] property will be renamed to the [rawSegments].
+     */
+    @Deprecated(
+        """
+        Path segments is deprecated. Please consider using rawSegments if you need this behavior instead.
+
+        This property will contain empty path segment at the beginning for the segment for the url with hostname,
+        and empty path segment at the end for the urls with trailing slash.
+        
+        If you need path segments 
+        """,
+        replaceWith = ReplaceWith("segments")
+    )
+    public val pathSegments: List<String> = pathSegments
+
+    /**
+     * A list containing the segments of the URL path.
+     *
+     * This property was designed to be distinguishing between an absolute and relative path,
+     * so it will have an empty segment at the beginning for URLs with a hostname
+     * and an empty segment at the end for URLs with a trailing slash.
+     *
+     * ```kotlin
+     * val fullUrl = Url("http://ktor.io/docs/")
+     * fullUrl.rawSegments ==  listOf("", "docs", "")
+     *
+     * val absolute = Url("/docs/")
+     * absolute.rawSegments ==  listOf("", "docs", "")
+     *
+     * val relative = Url("docs")
+     * relative.rawSegments = listOf("docs")
+     * ```
+     *
+     * It may be not comfortable if you're working only with full urls.
+     * If you don't need empty segments specific behavior, please consider using [segments] property:
+     *
+     * ```kotlin
+     * val fullUrl = Url("http://ktor.io/docs/")
+     * fullUrl.segments ==  listOf("docs")
+     *
+     * val absolute = Url("/docs/")
+     * absolute.segments ==  listOf("docs")
+     *
+     * val relative = Url("docs")
+     * relative.segments = listOf("docs")
+     * ```
+     */
+    public val rawSegments: List<String> = pathSegments
+
+    /**
+     * A list of path segments derived from the URL, excluding any leading
+     * and trailing empty segments.
+     *
+     * ```kotlin
+     * val fullUrl = Url("http://ktor.io/docs/")
+     * fullUrl.segments ==  listOf("docs")
+     *
+     * val absolute = Url("/docs/")
+     * absolute.segments ==  listOf("docs")
+     * val relative = Url("docs")
+     * relative.segments = listOf("docs")
+     * ```
+     *
+     * If you need to check for trailing slash and relative/absolute paths, please check the [rawSegments] property.
+     **/
+    public val segments: List<String> by lazy {
+        val start = if (pathSegments.isEmpty() || !pathSegments.first().isEmpty()) 0 else 1
+        val end = if (pathSegments.last().isEmpty()) pathSegments.lastIndex else pathSegments.lastIndex + 1
```

Comment:
`.last()` can throw `NoSuchElementException` if the list is empty, for example for `Url("https://ktor.io")`